python/ray/serve/deployment_state.py

```diff
@@ -357,48 +360,110 @@ def check_stopped(self) -> bool:
+            # There is no outstanding health check.
+            return None, False
 
-        healthy = True
+        health_check_succeeded: Optional[bool] = None
+        replica_actor_crashed: bool = False
         ready, pending = ray.wait([self._health_check_ref], timeout=0)
         if len(ready) == 1:
             # Object ref is ready, ray.get it to check for exceptions.
             try:
                 ray.get(self._health_check_ref)
-
-                # Health check succeeded. Kick off another and reset the timer
-                # if it's been long enough since the last health check. Add
-                # some randomness to avoid synchronizing across all replicas.
-                healthy = True
-                time_since_last = time.time() - self._last_health_check_time
-                randomized_period = self._health_check_period_s * random.uniform(
-                    0.9, 1.1
-                )
-                if time_since_last > randomized_period:
-                    self._last_health_check_time = time.time()
-                    self._health_check_ref = (
-                        self._actor_handle.check_health.remote()  # noqa: E501
-                    )
+                # Health check succeeded without exception.
+                health_check_succeeded = True
+                self._health_check_ref = None
+            except RayActorError:
+                # Health check failed due to actor crashing.
+                logger.info(f"Actor for replica {self._replica_tag} crashed.")
+                health_check_succeeded = False
+                self._health_check_ref = None
+                replica_actor_crashed = True
             except RayError as e:
-                logger.info(
-                    f"Health check for replica {self._replica_tag} " f"failed: {e}"
-                )
-                healthy = False
+                # Health check failed due to application-level exception.
+                logger.info(f"Health check for replica {self._replica_tag} failed: {e}")
+                health_check_succeeded = False
+                self._health_check_ref = None
         elif time.time() - self._last_health_check_time > self._health_check_timeout_s:
+            # Health check hasn't returned and the timeout is up, consider it failed.
             logger.info(
                 "Didn't receive health check response for replica "
                 f"{self._replica_tag} after "
                 f"{self._health_check_timeout_s}s, marking it unhealthy."
             )
-            healthy = False
+            health_check_succeeded = False
+            self._health_check_ref = None
         else:
-            healthy = True
+            # Health check hasn't returned and the timeout isn't up yet.
+            health_check_succeeded = None
+
+        return health_check_succeeded, replica_actor_crashed
 
-        return healthy
+    def _should_start_new_health_check(self) -> bool:
+        """TODO"""
```

Comment:
Is this still TODO?

Reply:
ah yeah I need to update docstrings, let me do that